### PR TITLE
feat(react-router): implement hash change "Scroll Into View" option

### PR DESCRIPTION
**Pull request by joshuaslate**
I am working on an API documentation page that has a number of anchor links. The sidebar contains `Link` elements that expand/collapse sections based on if its hash is active. It works like a table of contents. I also update the hash based on the user's scroll position. If they've scrolled past an anchor, it updates the hash. The issue I was hitting is that updating the hash with `navigate` or a number of other methods actually runs `scrollIntoView`, which takes over the user's scrolling in these cases.

The change I'm proposing is to add an option to `navigate` and `Link` that will allow the consumer to choose to opt out of the default `scrollIntoView` behavior, or even allow them to customize the scroll behavior if they would like to (e.g., for smooth scrolling instead of instant).

Discussed in Discord here: https://discord.com/channels/719702312431386674/1316559407919796334

Another relevant discussion: https://github.com/TanStack/router/discussions/1155

### CODE REVIEW

File: docs/framework/react/api/router/RouterType.md
Comment (schiller-manuel) [inline docs/framework/react/api/router/RouterType.md:116]: we should also add a `defaultHashChangeScrollIntoView` (or whatever this thing will be named finally) into router (defaults to `true`)

File: examples/react/basic/src/main.tsx
Comment (schiller-manuel) [inline examples/react/basic/src/main.tsx:89]: move this into `e2e/react-router/basic-file-based` and add a new `scrollIntoView.spec.ts` test (test name just a proposal)

File: packages/react-router/src/router.ts
Comment (schiller-manuel) [inline packages/react-router/src/router.ts:720]: remove this

File: docs/framework/react/api/router/RouterType.md
Comment (joshuaslate) [reply docs/framework/react/api/router/RouterType.md:116]: I added it to the router options and implemented it as I saw some of the other default options. Thanks!

File: examples/react/basic/src/main.tsx
Comment (joshuaslate) [reply examples/react/basic/src/main.tsx]: Done, thanks!

File: packages/react-router/src/router.ts
Comment (joshuaslate) [reply packages/react-router/src/router.ts]: Done, thanks!

File: e2e/react-router/basic-file-based/tests/scrollIntoView.spec.ts
Comment (schiller-manuel) [inline e2e/react-router/basic-file-based/tests/scrollIntoView.spec.ts:12]: please use getByTestId instead of getByRole (this needs a `data-testId` prop on the element you want to find

Comment (joshuaslate) [reply e2e/react-router/basic-file-based/tests/scrollIntoView.spec.ts]: Updated, thanks.

**Comment by schiller-manuel** [comment]
this looks great. let's wait for @SeanCassiere to review

**Review by schiller-manuel — APPROVED** [review]
LGTM

File: packages/react-router/src/router.ts
Comment (SeanCassiere) [inline packages/react-router/src/router.ts:319]: The MDN link should also be placed in the markdown files.

- RouterOptionsType.md
- RouterType.md

File: e2e/react-router/basic-file-based/tests/scrollIntoView.spec.ts
Comment (SeanCassiere) [inline e2e/react-router/basic-file-based/tests/scrollIntoView.spec.ts:1]: Could we rename the file to `scroll-into-view.spec.ts`

File: e2e/react-router/basic-file-based/src/routes/anchor.tsx
Comment (SeanCassiere) [inline e2e/react-router/basic-file-based/src/routes/anchor.tsx:121]: Please use `satisfies` here.

**Review by SeanCassiere — APPROVED** [review]
I'm not at home at the moment, so I'll have to check the sandbox locally at home tomorrow.

I'd also like us to think a bit more on the property name. The name `hashChangeScrollIntoView` just feels a bit off. Maybe even `hashScrollIntoView`?

Approving for now, but I'd like to get a chance to play with the sandbox before merging in.

**Comment by joshuaslate** [comment]
I believe I've addressed all of the feedback that's come in so far, please let me know if anything else comes up. I added the `/anchor` route to the `basic-file-based` example for easy manual testing in addition to the e2e tests.

File: packages/react-router/src/router.ts
Comment (SeanCassiere) [inline packages/react-router/src/router.ts:1850]: Would this (the `?? true`) break any existing apps?

Comment (joshuaslate) [reply packages/react-router/src/router.ts:1850]: These options are new, so the value will default to true if they aren't provided.

I read on MDN for scrollIntoView that true is the default argument value, so passing true by default should maintain the behavior that exists today. I don't believe this is a breaking change.

I'm not at my computer right now, but I believe I added a test case and Link on the example that use the default behavior, and it seems to function the same as the released version does for me.

**Review by SeanCassiere — APPROVED** [review]
Tried out the sandbox and it looks good.

@schiller-manuel OK to merge from my side.

**Comment by joshuaslate** [comment]
I was investigating the test failure from scroll restoration and found that if browser (not router)-controlled navigation happens, the `router.state.location.state.__hashScrollIntoViewOptions` option isn't being set, because it doesn't run through `commitLocation`.

I could add another nullish coalescing operator to set the value if it's undefined, like so:
```
if (typeof document !== 'undefined' && (document as any).querySelector) {
        const hashScrollIntoViewOptions =
          router.state.location.state.__hashScrollIntoViewOptions ?? true

        if (hashScrollIntoViewOptions && router.state.location.hash !== '') {
          const el = document.getElementById(router.state.location.hash)
          if (el) {
            el.scrollIntoView(hashScrollIntoViewOptions)
          }
        }
      }
```

This fixes the test and is working for me in the scroll restoration sandbox.

@SeanCassiere, does that sound ok to you?

**Comment by SeanCassiere** [comment]
> @SeanCassiere, does that sound ok to you?

Worth giving it a try.
I'll also be investigating what's going on here.

**Review by SeanCassiere — APPROVED** [review]

**Comment by SeanCassiere** [comment]
cc @schiller-manuel

**Pull request merged**
